uvm/plan9.go

```diff
@@ -0,0 +1,129 @@
+package uvm
+
+import (
+	"fmt"
+	//"path/filepath"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/Microsoft/hcsshim/internal/wclayer"
+	"github.com/Microsoft/hcsshim/uvm/lcowhostedsettings"
+	"github.com/sirupsen/logrus"
+)
+
+// AddPlan9 adds a Plan9 share to a utility VM. Each Plan9 share is ref-counted and
+// only added if it isn't already.
+func (uvm *UtilityVM) AddPlan9(hostPath string, uvmPath string, flags int32) error {
+	if uvm.operatingSystem != "linux" {
+		return errNotSupported
+	}
+	if uvmPath == "" {
+		return fmt.Errorf("uvmPath must be passed to AddPlan9")
+	}
+
+	hostPath = strings.ToLower(hostPath)
+	logrus.Debugf("uvm::AddPlan9 %s %s %d id:%s", hostPath, uvmPath, flags, uvm.id)
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+	if uvm.plan9Shares == nil {
+		uvm.plan9Shares = make(map[string]plan9Info)
+	}
+	if _, ok := uvm.plan9Shares[hostPath]; !ok {
+		guid, err := wclayer.NameToGuid(hostPath) // TODO: We should hash the full hostpath on VSMB too
+		if err != nil {
+			logrus.Debugf("Failed NamedToGuid", err)
+			return err
+		}
+
+		uvm.plan9PortCounter++ // TODO: This is temporary. Each share currently requires a unique port in HCS and GCS. This will change so we use a single port (9999)
+
+		modification := &schema2.ModifySettingsRequestV2{
+			ResourceType: schema2.ResourceTypePlan9Share,
+			RequestType:  schema2.RequestTypeAdd,
+			Settings: schema2.VirtualMachinesResourcesStoragePlan9ShareV2{
+				Name: guid.String(),
+				Path: hostPath,
+				Port: uvm.plan9PortCounter,
+			},
+			ResourceUri: fmt.Sprintf("virtualmachine/devices/plan9shares/%s", guid.String()),
+			HostedSettings: lcowhostedsettings.MappedDirectory{
+				MountPath: uvmPath,
+				Port:      uvm.plan9PortCounter,
+				ReadOnly:  (flags | schema2.VPlan9FlagReadOnly) == schema2.VPlan9FlagReadOnly,
+			},
+		}
+
+		if err := uvm.Modify(modification); err != nil {
+			return err
+		}
+		uvm.plan9Shares[hostPath] = plan9Info{
+			refCount: 1,
+			uvmPath:  uvmPath,
+			guid:     guid.String(),
+			port:     uvm.plan9PortCounter,
+		}
+	} else {
+		p9i := plan9Info{
```

Comment:
In network.go I have a simpler approach to updating the map, by using a pointer for the map's value type.